smdebug/tensorflow/keras.py

```diff
@@ -626,97 +626,157 @@ def new_add_weight(opt, *args, **kwargs):
+            ):
+                return grads
 
-    def record_tensor_value(self, tensor_name, tensor_value):
+            if self._get_collections_to_save_for_step():
+                for (g, v) in zip(grads, vars):
+                    layer = v.name.split(":")[0]
+                    self._save_for_tensor(
+                        tensor_name="gradients/" + layer + "Grad",
+                        tensor_value=g,
+                        check_before_write=True,
+                    )
+                    self._save_for_tensor(
+                        tensor_name="weights/" + v.name,
+                        tensor_value=v.value(),
+                        check_before_write=True,
+                    )
+
+            if not ((isinstance(loss, tf.Tensor)) and hasattr(loss, "numpy")):
+                return grads
+
+            self._add_metric(metric_name="loss", metric_value=loss)
+            if self._is_collection_being_saved_for_step(CollectionKeys.LOSSES):
+                self._initialize_writers(only_initialize_if_missing=True)
+                self._save_for_tensor("loss", loss, check_before_write=False)
+
+            return grads
+
+        return run
+
+    def _wrap_pop_tape(self, function):
         """
-        Used to manually set losses and metrics when GradientTape is
-        used in the training script
+        tape._pop_tape() is called at the end of a GradientTape execution.
+        Using this to export collections
         """
-        # To be used to save losses and metrics of type EagerTensor
-        if (
-            not ((isinstance(tensor_value, tf.Tensor)) and hasattr(tensor_value, "numpy"))
-        ) or self._is_not_supported():
-            return
 
-        self._add_metric(metric_name=tensor_name, metric_value=tensor_value)
-        if self._is_collection_being_saved_for_step(
-            CollectionKeys.METRICS
-        ) or self._is_collection_being_saved_for_step(CollectionKeys.LOSSES):
-            self._initialize_writers(only_initialize_if_missing=True)
-            self._save_for_tensor(tensor_name, tensor_value, check_before_write=False)
+        @functools.wraps(function)
+        def run(*args, **kwargs):
+            function(*args, **kwargs)
+            if self._is_not_supported():
+                return
 
-    def forward_post_hook(self):
-        """
-        Export collections and model at the end of a step.
+            if self._exported_collections is False:
+                # in keras, these collections change when mode changes
+                # but rest of the project isn't yet capable of handling this
+                # this means that collections like outputs, or other collections with intermediate tensors
+                # will only have tensor names from first mode
+
+                # this means sometimes collections will be exported after 1 step
+                self.export_collections()
+                self._exported_collections = True
```

Comment:
Some hooks use the variable `self.exported_collections` and some use `self._exported_collections`. Maybe not within the scope of this PR, but that should be fixed, and setting the marker boolean to True should occur within the method, not outside of it.